### PR TITLE
H207: weight interpolation H112 EP13 ↔ H183 EP13

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

H112 EP13 has the basin-intact slope (val 6.1358%, test 6.752%, WSS_x slope −0.215pp). H183 EP13 has the program val SOTA (val 6.039%) but basin-disrupted terminal (test 6.829%, WSS_x slope +0.064pp). These are two checkpoints on the **same problem manifold**.

**Hypothesis**: Linear interpolation of the two state_dicts at intermediate alpha produces a checkpoint with PARTIAL val gain from H183 AND PARTIAL basin integrity from H112. If a sweet spot exists where val < 6.1358 AND WSS_x slope < −0.020pp AND test_WSS < 6.752 → SOTA candidate at zero training cost.

```
W_alpha = (1 − alpha) × W_H112 + alpha × W_H183
```

- alpha = 0.0 → pure H112 (current SOTA)
- alpha = 1.0 → pure H183 (val SOTA, test fails)
- alpha ∈ (0, 1) → unexplored manifold

## Instructions

**This is a pure eval sprint — NO training. No source file modifications.**

### Step 1: Locate both EP13 best checkpoints
- H112: W&B run `u9ue2ryb`, artifact `model-*-u9ue2ryb:v0` (alias `best`)
- H183: W&B run `5k58uzqc`, artifact `model-fern-h183-mirror-aug-tau-y-3p0-compound-5k58uzqc:v0` (alias `best`)
- Download both to local disk

### Step 2: Linear interpolation pipeline

```python
import torch

w_h112 = torch.load("h112_ep13.pt")  # state_dict (or full ckpt — extract state_dict if nested)
w_h183 = torch.load("h183_ep13.pt")

# IMPORTANT: ensure both have IDENTICAL keys and tensor shapes.
# H112 and H183 used same architecture (depth=5, hidden=512, etc.), so this should hold.
# If keys differ (e.g. EMA wrapper artifacts), strip prefixes consistently before interpolation.

for alpha in [0.0, 0.25, 0.5, 0.75, 1.0]:
    w_alpha = {k: (1 - alpha) * w_h112[k] + alpha * w_h183[k] for k in w_h112}
    torch.save(w_alpha, f"interp_alpha_{alpha}.pt")
```

### Step 3: Eval each interpolated checkpoint (DDP=8 GPUs)

```bash
for alpha in 0.0 0.25 0.5 0.75 1.0; do
  python -m torch.distributed.run --nproc-per-node=8 \
    train.py --eval_only \
    --checkpoint interp_alpha_${alpha}.pt \
    --wandb_group h207-askeladd-interp-h112-h183 \
    --run_name h207-alpha-${alpha}
done
```

### Step 4: Report

Table per alpha:
- val_abupt, val_WSS, val_WSS_x, val_WSS_y, val_WSS_z, val_VP, val_SP
- test_WSS, test_WSS_x, test_WSS_y, test_WSS_z, test_VP, test_SP
- WSS_x slope sign (key basin diagnostic)
- val→test slope on WSS aggregate

### Step 5: SOTA candidate check

Criterion: val_abupt < 6.1358% AND test_WSS ≤ 6.752% AND test_VP ≤ 3.421% AND test_SP ≤ 3.577% AND WSS_x slope ≤ −0.020pp.

If any alpha passes all 5 gates → SOTA candidate. Report the SOTA-candidate alpha + per-channel decomposition. Tag advisor for merge.

**Bonus arm — find optimum alpha**:
After the initial 5-point sweep, if alpha=0.25 or 0.5 looks promising (basin intact + val gain), do a refinement sweep at alpha=0.1, 0.2, 0.3, 0.4. Total ~9 evals ≈ 2.5h on 8 GPUs.

## Baseline

Current best (H112, PR #1283, W&B run `u9ue2ryb`):
- val_abupt: **6.1358%**
- test_WSS: **6.752%**
- test_VP: **3.421%**
- test_SP: **3.695%** (strict gate ≤ 3.577%)

All 4 gates + WSS_x basin must pass for merge.

## Coordination

- fern H208 = same logic, H112 ↔ H190 interpolation
- tanjiro H210 = cross-recipe SWA of all 5 EP13 best artifacts (the full-cohort consensus variant)
- These 3 weight-arithmetic experiments together test whether the val-SOTA recipes can be partially recovered via parameter-space averaging

## Background

The original H198 (cohort-wide checkpoint search) was blocked by the same checkpoint-persistence issue that blocked H204 — mid-EP EMAs were never saved. H207 PIVOTS to weight interpolation between TWO EP13 best artifacts, which is fully executable.

Expected runtime: ~1.5-2h on 8 GPUs (5-9 evals × ~20-25min each).

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT askeladd:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["69ul5jj7","u4mttmbd","5w1mo8zj","ynkwv226","vuj4sqpi"],"primary_metric":{"name":"full_val_primary/abupt_axis_mean_rel_l2_pct","value":6.0388},"test_metric":{"name":"test_primary/abupt_axis_mean_rel_l2_pct","value":5.8391}}

## Results — H207 weight interpolation H112 EP13 ↔ H183 EP13

**Hypothesis REJECTED.** The two EP13 checkpoints are NOT linearly mode-connected. Every intermediate alpha collapses catastrophically; no SOTA candidate exists on the linear interpolation manifold.

### Full 5-point sweep

| alpha | run_id | val_abupt | test_abupt | test_WSS | test_WSS_x | test_WSS_y | test_WSS_z | test_VP | test_SP | WSS_x slope (val→test) |
|---|---|---|---|---|---|---|---|---|---|---|
| 0.0 (H112) | 69ul5jj7 | **6.1358** | 5.8391 | **6.7523** | 5.9990 | 7.3604 | 8.7200 | **3.4213** | 3.6948 | −0.093pp |
| 0.25 | u4mttmbd | 66.6014 | 66.7094 | 72.03 | 65.83 | 83.30 | 80.43 | 49.86 | 54.13 | +0.84pp |
| 0.5 | 5w1mo8zj | 88.7724 | 88.8888 | 91.92 | 86.22 | 101.81 | 101.01 | 79.13 | 76.27 | +0.34pp |
| 0.75 | ynkwv226 | 62.5444 | 63.1741 | 66.93 | 59.58 | 81.18 | 74.69 | 50.11 | 50.31 | +0.75pp |
| 1.0 (H183) | vuj4sqpi | **6.0388** | 5.9136 | 6.8287 | 6.1121 | 7.3295 | 8.8154 | 3.5277 | 3.7833 | +0.064pp |

### Endpoint sanity check ✅

The endpoints exactly reproduce the published baselines, confirming the eval pipeline is correct:
- alpha=0.0 ≡ H112 (PR #1283): val 6.1358 ✓ (PR baseline 6.1358), test_WSS 6.7523 ✓ (6.752), test_VP 3.4213 ✓ (3.421), test_SP 3.6948 ✓ (3.695)
- alpha=1.0 ≡ H183 EP13: val 6.0388 ✓ (6.039), test_WSS 6.8287 ✓ (6.829), WSS_x slope +0.064pp ✓ (+0.064)

### SOTA-candidate gate check

Required: val_abupt < 6.1358 AND test_WSS ≤ 6.752 AND test_VP ≤ 3.421 AND test_SP ≤ 3.577 AND WSS_x slope ≤ −0.020pp.

| alpha | val gate | test_WSS | test_VP | test_SP | WSS_x slope | PASS? |
|---|---|---|---|---|---|---|
| 0.0 | = baseline | = baseline | = baseline | 3.695 (fails strict SP) | −0.093 ✓ | **NO** |
| 0.25 | FAIL (66.6) | FAIL | FAIL | FAIL | FAIL | **NO** |
| 0.5 | FAIL (88.8) | FAIL | FAIL | FAIL | FAIL | **NO** |
| 0.75 | FAIL (62.5) | FAIL | FAIL | FAIL | FAIL | **NO** |
| 1.0 | 6.04 ✓ | 6.829 (fails 6.752) | 3.528 (fails 3.421) | 3.783 (fails 3.577) | +0.064 ✗ | **NO** |

**No alpha passes all 5 gates. No SOTA candidate.**

### Bonus-arm decision

Bonus arm triggers only if alpha=0.25 or 0.5 looks promising (basin intact + val gain). Both are catastrophically broken (val_abupt 66.6 and 88.8 vs 6.04 endpoints), so the bonus arm is **NOT run** per PR instructions. The basin walls between H112 and H183 are extremely sharp — a 10× val collapse over a 0.25 alpha step indicates we're outside the loss basin almost immediately.

### What happened

The hypothesis was that H112 and H183 EP13 sit on the same loss basin (just at different points), so a convex combination could blend H112's basin-intact slope with H183's better val. The data says the opposite: **they are in two distinct basins**.

Three pieces of evidence:

1. **Catastrophic interpolation collapse.** Even alpha=0.25 (75% H112 + 25% H183) collapses val_abupt from 6.14 → 66.6 — an order-of-magnitude jump. Pure linear mode-connectivity (à la Frankle et al. 2020, Entezari et al. 2021) fails here.

2. **Asymmetric collapse profile.** alpha=0.25 → val 66.6, alpha=0.5 → 88.8, alpha=0.75 → 62.5. The maximum is at the midpoint (deepest mountain pass), and the H183-side wall (0.75) is slightly less brutal than the H112-side wall (0.25). The two basins have different curvature/orientation.

3. **All three target families collapse together.** Surface pressure, wall shear (all three axes), and volume pressure all degrade by the same ~10×–15× factor at intermediate alpha. This is a global model failure, not a single-head pathology. The shared backbone weights are not in a flat region between the two solutions.

This is consistent with prior empirical findings that early-stopped checkpoints from independent training trajectories typically require permutation alignment (Ainsworth et al. *Git Re-Basin*, 2023) before linear interpolation works — and even then, only when the architecture has enough symmetry.

### Exact command used

```bash
cd /workspace/senpai/target
for alpha in 0.0 0.25 0.5 0.75 1.0; do
  python interp_eval.py --alphas $alpha \
    --wandb-name-prefix h207-alpha \
    --wandb-group h207-askeladd-interp-h112-h183 \
    2>&1 | tee run_logs/h207/alpha_${alpha}.log
done
```

(Single-process eval per alpha on one GPU — `interp_eval.py` reuses `ensemble_eval.evaluate_ensemble_split` for full val + test in one pass; matches the same code path used to reproduce the baselines.)

### Peak memory & wall-clock

- Peak GPU memory: **9.1 GiB** (well under headroom; single-GPU eval).
- Wall-clock per alpha: ~1210s (20.2 min) full val + full test.
- Total sweep: ~101 min for 5 alphas.

### W&B runs

| alpha | run_id | URL |
|---|---|---|
| 0.0 | 69ul5jj7 | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/69ul5jj7 |
| 0.25 | u4mttmbd | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/u4mttmbd |
| 0.5 | 5w1mo8zj | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/5w1mo8zj |
| 0.75 | ynkwv226 | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/ynkwv226 |
| 1.0 | vuj4sqpi | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/vuj4sqpi |

All runs tagged `h207`, `interp`, `askeladd`, `eval-only`; grouped under `h207-askeladd-interp-h112-h183`.

### Suggested follow-ups

1. **Permutation-aligned interpolation (Git Re-Basin).** The deep collapse strongly suggests neuron permutation symmetry is breaking interpolation. Apply weight matching (Ainsworth 2023) to H183 → H112 ordering before interpolation. Cheap to try, could legitimately reveal linear connectivity if the basins are *the same up to permutation*. ~1 day of work + a sweep.

2. **Output-logit averaging instead of weight averaging.** If weight-space interpolation collapses, function-space ensembling typically still works: at eval time, predict with both models and average the *outputs* (surface_preds and volume_preds). Same idea as tanjiro H210 SWA but in function space. Cheap eval-only follow-up; potentially gets the "best of both" without the basin problem.

3. **Implications for H208 (fern, H112↔H190) and H210 (tanjiro, 5-way EP13 SWA).** Both also assume mode connectivity. H208 is at risk of the same failure if H190 ≠ H112 basin. H210's broader SWA over 5 independently-trained recipes is very likely to suffer the same collapse. The advisor should consider whether to redirect those experiments toward Git Re-Basin or function-space ensembling, or wait for H208 result to confirm the pattern.

4. **The H183 basin-disrupted terminal (test_WSS 6.829 vs val 6.039) is real.** The "val SOTA but test fail" pattern is reproduced exactly by alpha=1.0. This rules out evaluation noise as the cause — H183 genuinely overfits to val. Future improvements need basin-intact training (H112-style), not just better val numbers.

5. **One artifact note.** Code added: `interp_eval.py` (new, ~270 LOC, eval-only, reuses `ensemble_eval` infrastructure). Did not modify any protected files. Will commit on submit.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1380 because ADVISOR: **H207 closed — CLEAN NEGATIVE. Finding P (basins not linearly connected) banked.**

Excellent execution. Endpoint sanity at alpha=0 (val 6.1358 ✓ test_WSS 6.7523 ✓ to ≤0.0001pp) and alpha=1 (val 6.0388 ✓ test_WSS 6.8287 ✓) confirms the eval pipeline. The catastrophic collapse at every intermediate alpha is a clean signal.

### Finding P (program-permanent): H112 ↔ H183 basins are NOT linearly mode-connected

| alpha | val_abupt | test_WSS | functional? |
|---|---:|---:|---|
| 0.0 (H112) | 6.1358 | 6.752 | ✓ canonical |
| 0.25 | 66.60 | 72.03 | DESTROYED (10× degradation in 1 alpha step) |
| 0.5 | 88.77 | 91.92 | DESTROYED (deepest mountain pass) |
| 0.75 | 62.54 | 66.93 | DESTROYED (slightly less brutal H183 side) |
| 1.0 (H183) | 6.0388 | 6.829 | ✓ canonical |

The asymmetric collapse profile (0.25 wall slightly steeper than 0.75 wall) hints at different basin curvature/orientation. Per your analysis, this is consistent with neuron-permutation symmetry breaking interpolation (Ainsworth et al. 2022 *Git Re-Basin*, Entezari et al. 2021).

**This cross-validates Finding O (tanjiro H210 5-way SWA)**: independently-trained Transformers settle in permutation-equivalent but parameter-distinct basins. The 2-way collapse at alpha=0.25 is the predicted consequence.

### Implications for fleet

- **fern H208 (H112↔H190 interp)**: same collapse expected. Already running, will validate.
- **tanjiro H213 (block-wise splice)**: this is the right alternative. Block-splicing keeps each block's internal parameters coherent — no permutation conflict within a block. Watch his results.
- **TTA arm (H192/H206/H209/H211)**: orthogonal mechanism (no weight averaging), unaffected by Finding O/P.

### Git Re-Basin not feasible in current window

Your follow-up suggestion #1 (permutation alignment before interpolation) is correct but ~1 day of work. We have ~7h. Tabling.

### Next assignment: H214 — sub-alpha sweep near H112 endpoint

Probe the H112 basin **radius** in the H183 direction with finer granularity: alpha ∈ {0.005, 0.01, 0.025, 0.05, 0.1, 0.15, 0.2}. At what minimal alpha does the model stop being functional? If alpha=0.05 still preserves H112's basin → tiny workable region exists. If alpha=0.005 already collapses → completely disjoint, no linear blend possible at any meaningful weight.

Branch incoming. Closing this PR. Result logged..
